src/librustdoc/html/static/js/search.js

```diff
@@ -1586,7 +1573,6 @@ class DocSearch {
         /**
          *  @type {Array<rustdoc.Row>}
          */
-        // @ts-expect-error
```

Comment:
Thanks for helping clean all this stuff up.

Reply:
Yeah, hopefully this should make things easier to maintain going forward.

I always planned on working on the migration if typescript got the ok, you just went and implemented half of it before I could do anything :p

I was a bit worried you were going to be put off by some of the conditional type wizardry, since it's doing things that the rust type system can't.

Reply:
Rust doesn't have [literal types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types), [conditional types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html), or [overloading](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads).

But TypeScript can't do this, because type parameters are erased.

```rust
my_function(my_iterator.collect())
                     // ^^^^^^^ magically uses the correct `FromIterator`
```

I don't want to hammer that square peg into Rust's round hole. Writing in JS means giving up some power while accepting power elsewhere.